plugin/lighthouse/handler.go

```diff
@@ -0,0 +1,61 @@
+package lighthouse
+
+import (
+	"context"
+	"net"
+	"strings"
+
+	"github.com/coredns/coredns/plugin"
+	"github.com/coredns/coredns/plugin/pkg/nonwriter"
+	"github.com/coredns/coredns/request"
+
+	"github.com/miekg/dns"
+)
+
+// ServeDNS implements the plugin.Handler interface.
+func (lh *Lighthouse) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
+	state := request.Request{W: w, Req: r}
+
+	a := new(dns.Msg)
+	a.SetReply(r)
+	a.Authoritative = true
+
+	services := lh.SvcsMap
+	svcName := strings.Split(state.QName(), ".")[0]
+	ip := services[svcName]
+	if ip == "" {
+		return onFailure(lh, ctx, w, r)
+	}
+	rr := new(dns.A)
+
+	if state.Family() == 1 {
+		// IPv4 query
+		rr.Hdr = dns.RR_Header{Name: state.QName(), Rrtype: dns.TypeA, Class: state.QClass()}
+		rr.A = net.ParseIP(ip).To4()
+	} else {
+		log.Debugf("IPv6 queries not supported yet")
+		return onFailure(lh, ctx, w, r)
+	}
+
+	a.Answer = []dns.RR{rr}
+
+	log.Debugf("Responding to query with '%s'", a.Answer)
+	if w.WriteMsg(a) != nil {
+		log.Debugf("Failed to writeMsg")
+		return onFailure(lh, ctx, w, r)
+	}
+
+	return dns.RcodeSuccess, nil
+}
+
+func onFailure(lh *Lighthouse, ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
+	// Use a non-writer so we don't write NXDOMAIN to client
+	nw := nonwriter.New(w)
```

Comment:
Is this a pattern copied from other plugins? It seems weird to chain to another plugin but not give it a usable writer!

Reply:
This one's from reference plugin. I'll probably visit it later once get a better handle on DNS protocol and how lighthouse fits with other plugins. Other plugins will use their own writer [all do in their ServeDNS, I think this is just to avoid *us* writing NXDOMAIN to client.

Reply:
Tried out a few things and you're correct. The use case for non writer is different from what we're doing over here. Corrected this.